# Summary of EDA

## 1: Importing Libraries and Loading Data

In [53]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import sqlite3
import datetime as dt

In [7]:
# Settings
%matplotlib inline
pd.set_option("display.max_columns", None)

In [4]:
# Connecting to database
conn = sqlite3.connect(r"./data/lung_cancer.db")

In [34]:
# Reading the data
MASTER_DF = pd.read_sql_query("SELECT * FROM lung_cancer", conn)

## 2: Exploring the Data

### 2.1: Brief General Overview

In [35]:
MASTER_DF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10348 entries, 0 to 10347
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   ID                      10348 non-null  float64
 1   Age                     10348 non-null  int64  
 2   Gender                  10348 non-null  object 
 3   COPD History            9236 non-null   object 
 4   Genetic Markers         10348 non-null  object 
 5   Air Pollution Exposure  10345 non-null  object 
 6   Last Weight             10348 non-null  int64  
 7   Current Weight          10348 non-null  int64  
 8   Start Smoking           10348 non-null  object 
 9   Stop Smoking            10348 non-null  object 
 10  Taken Bronchodilators   9287 non-null   object 
 11  Frequency of Tiredness  10348 non-null  object 
 12  Dominant Hand           10348 non-null  object 
 13  Lung Cancer Occurrence  10348 non-null  int64  
dtypes: float64(1), int64(4), object(9)
mem

In [68]:
MASTER_DF.describe()

ID          Age  Last Weight  Current Weight  \
count  9.998000e+03  9998.000000  9998.000000     9998.000000   
mean   1.152093e+18    64.059412    72.383877       69.407982   
std    6.673658e+17    14.253842    16.135922       16.878405   
min    4.137510e+13    18.000000    45.000000       36.000000   
25%    5.662028e+17    54.000000    59.000000       56.000000   
50%    1.154280e+18    64.000000    72.000000       69.000000   
75%    1.722367e+18    74.000000    86.000000       82.000000   
max    2.305810e+18   120.000000   100.000000      115.000000   

       Lung Cancer Occurrence  
count             9998.000000  
mean                 0.544009  
std                  0.498084  
min                  0.000000  
25%                  0.000000  
50%                  1.000000  
75%                  1.000000  
max                  1.000000

In [36]:
MASTER_DF.head()

ID  Age Gender COPD History Genetic Markers  \
0  1.468154e+16   42   Male           No     Not Present   
1  1.271684e+18   72   Male          Yes     Not Present   
2  1.708068e+18   65   Male          Yes     Not Present   
3  3.555829e+17   67   Male           No     Not Present   
4  1.684598e+18   54   Male          Yes     Not Present   

  Air Pollution Exposure  Last Weight  Current Weight Start Smoking  \
0                 Medium           73              69          2015   
1                   High           50              42          2006   
2                    Low           65              75          1998   
3                   High           69              58          2011   
4                 Medium           46              40          1994   

    Stop Smoking Taken Bronchodilators Frequency of Tiredness Dominant Hand  \
0           2020                    No             None / Low         Right   
1           2012                   Yes                 Medium         Right   
2  Still Smoking                   Yes             None / Low         Right   
3           2014                    No             None / Low         Right   
4           2017                   Yes             None / Low         Right   

   Lung Cancer Occurrence  
0                       1  
1                       1  
2                       0  
3                       1  
4                       0

### 2.2: Duplicated Data

In [37]:
# Checking for duplicates
MASTER_DF.duplicated().any()

True

In [38]:
MASTER_DF[MASTER_DF.duplicated() == True]

ID  Age  Gender COPD History Genetic Markers  \
9998   1.186573e+18   49  Female          Yes         Present   
9999   1.948079e+18   80    Male           No     Not Present   
10000  1.421010e+18   56  Female           No     Not Present   
10001  1.630550e+18   89  Female           No         Present   
10002  3.915065e+17   51    Male           No     Not Present   
...             ...  ...     ...          ...             ...   
10343  6.202150e+17   47    Male          Yes         Present   
10344  1.157407e+18   59  Female          Yes     Not Present   
10345  1.488014e+18   34    Male           No     Not Present   
10346  3.987035e+17   69    Male           No     Not Present   
10347  1.297767e+18   60    MALE         None     Not Present   

      Air Pollution Exposure  Last Weight  Current Weight   Start Smoking  \
9998                  Medium           56              46            2013   
9999                     Low           92              95  Not Applicable   
10000                   High           45              50            1999   
10001                   High           84              87            1950   
10002                   High           83              88            2014   
...                      ...          ...             ...             ...   
10343                    Low           55              50            2009   
10344                    Low           80              81            2008   
10345                   High           65              57            2020   
10346                   High           81              75            2014   
10347                 Medium           82              75            1969   

         Stop Smoking Taken Bronchodilators Frequency of Tiredness  \
9998    Still Smoking                   Yes             None / Low   
9999   Not Applicable                    No                   High   
10000            2021                    No             None / Low   
10001            2020                    No                 Medium   
10002   Still Smoking                    No             None / Low   
...               ...                   ...                    ...   
10343   Still Smoking                   Yes                 Medium   
10344            2011                   Yes                 Medium   
10345   Still Smoking                    No                 Medium   
10346   Still Smoking                    No                 Medium   
10347            1982                    No             None / Low   

      Dominant Hand  Lung Cancer Occurrence  
9998           Left                       1  
9999          Right                       1  
10000         Right                       0  
10001         Right                       1  
10002         Right                       0  
...             ...                     ...  
10343         Right                       0  
10344     RightBoth                       0  
10345     RightBoth                       1  
10346         Right                       1  
10347         Right                       0  

[350 rows x 14 columns]

In [39]:
# Validating those rows are indeed duplicated
MASTER_DF[MASTER_DF.duplicated(keep=False) == True].sort_values("ID")

ID  Age  Gender COPD History Genetic Markers  \
1271   9.360174e+15   65    Male          Yes         Present   
10275  9.360174e+15   65    Male          Yes         Present   
10290  1.788814e+16   61    Male          Yes     Not Present   
658    1.788814e+16   61    Male          Yes     Not Present   
10198  1.998191e+16   65  Female          Yes         Present   
...             ...  ...     ...          ...             ...   
6908   2.285686e+18   69    Male         None     Not Present   
294    2.290146e+18   78    Male          Yes         Present   
10176  2.290146e+18   78    Male          Yes         Present   
10093  2.298460e+18   74    MALE          Yes     Not Present   
3480   2.298460e+18   74    MALE          Yes     Not Present   

      Air Pollution Exposure  Last Weight  Current Weight   Start Smoking  \
1271                     Low           85              84            1995   
10275                    Low           85              84            1995   
10290                   High           94              93  Not Applicable   
658                     High           94              93  Not Applicable   
10198                   High           81              72            1994   
...                      ...          ...             ...             ...   
6908                     Low           90              83            2015   
294                     High           45              46            1980   
10176                   High           45              46            1980   
10093                 Medium           91              95            2019   
3480                  Medium           91              95            2019   

         Stop Smoking Taken Bronchodilators Frequency of Tiredness  \
1271             2022                  None             None / Low   
10275            2022                  None             None / Low   
10290  Not Applicable                   Yes             None / Low   
658    Not Applicable                   Yes             None / Low   
10198   Still Smoking                   Yes                 Medium   
...               ...                   ...                    ...   
6908    Still Smoking                  None             None / Low   
294              2002                   Yes                   High   
10176            2002                   Yes                   High   
10093            2019                   Yes             None / Low   
3480             2019                   Yes             None / Low   

      Dominant Hand  Lung Cancer Occurrence  
1271      RightBoth                       0  
10275     RightBoth                       0  
10290         Right                       1  
658           Right                       1  
10198         Right                       1  
...             ...                     ...  
6908          Right                       0  
294           Right                       1  
10176         Right                       1  
10093     RightBoth                       0  
3480      RightBoth                       0  

[700 rows x 14 columns]

From above 2 output tables, we can identify that there are 350 duplicated rows that should be removed

In [64]:
# Removing duplicated rows and validation
print(f"MASTER_DF shape before removal: {MASTER_DF.shape}")
MASTER_DF.drop_duplicates(keep="first", inplace=True)
print(f"Any Duplicate Rows after removal: {MASTER_DF.duplicated().any()}")
print(f"MASTER_DF shape after removal: {MASTER_DF.shape}")

MASTER_DF shape before removal: (9998, 14)
Any Duplicate Rows after removal: False
MASTER_DF shape after removal: (9998, 14)


MASTER_DF shape from the info table above: (10348, 14)

MASTER_DF shape after removal of duplicated rows: (9998, 14)

350 rows are removed, validating the numbers of rows removed tally with the number of duplicated rows we discovered earlier.

### 2.3: Data Columns Values

In [63]:
for col in MASTER_DF.columns[1:]:
	print(f"Column: {col}")
	print(f"{MASTER_DF[col].unique()}\n")

Column: Age
[  42   72   65   67   54   49   57   50   68   48   56   52   77   37
   78   74   53   70   61   45   63   47   84   59   76   79   71   69
  -61   60   41   73  110   66   88   29   18   40  -69   51   28   33
   91   83   34  -64   64   55   81   87   43   26   86   31   36   80
  -46   89   96   82   85   97   62   90   58   22   92   75   44   21
  100  103   95   38  -42   35   46   39  -49   93  101  -76  -65  -53
  -41  -84  -71   32  -55   99  -51  -87   27  -52  106   98   25  -80
  -54  -37  105  -74  108   20   94  -86  -72  -45  -77  -93  -40  102
  -81  -78   30   24  -90  -50  -66  -95  -68  -38 -101  -44  -82  -60
  -47  -56  115  -57  -73  120  -58  -75   23  -62  104  -79  -89  -88
  116  -43  107  -59  -33  -48  -98  -30  -70  -99  -92  -36 -100  -63
  -83 -104  -91]

Column: Gender
['Male' 'MALE' 'Female' 'FEMALE' 'NAN']

Column: COPD History
['No' 'Yes' None]

Column: Genetic Markers
['Not Present' 'Present']

Column: Air Pollution Exposure
['Medium' '

From the brief run through of the values above, it can be immediately highlighted that certain columns have values that require further investigation/manipulation:

1. Age column having negative values
2. Gender column having similar values ("Male" VS "MALE", "Female" VS "FEMALE) recorded with different cases
3. Start Smoking/Stop Smoking columns are not numerical, have words like "Not Applicable" and "Still Smoking"
4. None, NAN values in multiple columns (Possible Missing Data)

#### Features Table Summary Breakdown

| Feature | Description | Type | Sub-Type | Initial Remarks |
| - | - | - | - | - |
| ID | Patient ID | Numerical | Identifier | Not meaningful as a feature as it only serves a patient identifier |
| Age | Age of the patient | Numerical | Discrete | Contains negative values |
| Gender | Gender of the patient | Categorical | Nominal | Need to format same strings to one representative format |
| COPD History | Whether the patient has a history of Chronic Obstructive Pulmonary Disease | Categorical | Nominal | NIL |
| Genetic Markers | Presence of any genetic markers known to increase the risk of lung cancer | Categorical | Nominal | NIL |
| Air Pollution Exposure | Level of air pollution exposure in the patient’s daily life | Categorical | Ordinal | NIL |
| Last Weight | Last officially recorded weight of patient | Numerical | Nominal | Might require feature engineering together with Current Weight |
| Current Weight | Current officially recorded weight of patient | Numerical | Nominal | Might require feature engineering together with Last Weight |
| Start Smoking | Year that the patient starts smoking | String | N.A | Might require feature engineering together with Stop Smoking to be useful |
| Stop Smoking | Year that the patient stops smoking | String | N.A | Might require faeture engineering together with Start Smoking to be useful |
| Taken Bronchodilators | Whether the patient is previously prescribed Bronchodilator medications | Categorical | Nominal | NIL |
| Frequency of Tiredness | Frequency of patient feeling tiredness in a day | Categorical | Ordinal | NIL |
| Dominant Hand | Dominant hand of the patient | Categorical | Nominal | NIL |
| Lung Cancer Occurrence | Whether the patient has lung cancer or not | Categorical | Nominal | Target Variable |

MASTER_DF.drop("ID", axis=1, inplace=True)
MASTER_DF.head(n=5)

In the following sub-segment of 2.3: Data Columns Exploration, I will look into each of the issues highlighted earlier and make appropriate corrections to those that can be corrected

#### 2.3.1: Age Column having negative values

Assumption(s): It is widely known that age cannot be of negative values, thus it is highly likely that the values are simply keyed wrongly and should just be the direct positive value

In [65]:
MASTER_DF["Age"] = [age if age >= 0 else -age for age in MASTER_DF["Age"]]
MASTER_DF["Age"].unique()

array([ 42,  72,  65,  67,  54,  49,  57,  50,  68,  48,  56,  52,  77,
        37,  78,  74,  53,  70,  61,  45,  63,  47,  84,  59,  76,  79,
        71,  69,  60,  41,  73, 110,  66,  88,  29,  18,  40,  51,  28,
        33,  91,  83,  34,  64,  55,  81,  87,  43,  26,  86,  31,  36,
        80,  46,  89,  96,  82,  85,  97,  62,  90,  58,  22,  92,  75,
        44,  21, 100, 103,  95,  38,  35,  39,  93, 101,  32,  99,  27,
       106,  98,  25, 105, 108,  20,  94, 102,  30,  24, 115, 120,  23,
       104, 116, 107], dtype=int64)

#### 2.3.2: Gender column having similar values ("Male" VS "MALE", "Female" VS "FEMALE) recorded with different cases

In [67]:
MASTER_DF["Gender"] = [gender.lower().strip() if not None else "" for gender in MASTER_DF["Gender"]]
MASTER_DF["Gender"].unique()

array(['male', 'female', 'nan'], dtype=object)

#### 2.3.3: None, NAN values in multiple columns (Possible Missing Data)

In [48]:
NA_VARS = MASTER_DF.columns[MASTER_DF.isnull().sum(axis=0) > 0].tolist()

In [49]:
MASTER_DF[NA_VARS].isnull().mean().sort_values(ascending=False)

COPD History              0.108122
Taken Bronchodilators     0.102721
Air Pollution Exposure    0.000300
dtype: float64

All are Categorical

In [ ]:
# TODO: Appropriate Imputation

### 2.4: Vizualizing the Data

TODO: TARGET VARIABLE, NUMERICAL Variables, CATEGORICAL Variables, Correlation Analysis

LOOK AT DISTRIBUTIONS, OUTLIERS, CORRELATION

## 3: Feature Engineering

TODO: Investigate Start/Stop Smoking, Previous/Current Weight